lib/features/modeling/cmd/UpdateFlowNodeRefsHandler.js

```diff
@@ -42,6 +42,10 @@ UpdateFlowNodeRefsHandler.prototype.computeUpdates = function(flowNodeShapes, la
 
   var allFlowNodeShapes = [];
 
+  function getNodeKey(node) {
```

Comment:
This hack will only work on the assumption that the type always changes during replacement (which is not necessarily true). Did you consider an alternative solution approach that works without it?

Reply:
The bug does not occur when I change elements using the replace pad (for example from one intermediate throw event to another one), it only occurs on Attach/Detach event behavior. So I assumed this would work. Is there a case where this would break? In any case I'll investigate why the bug does not occur on replace, but occurs on Attach/Detach behavior.

Reply:
The bug does not occur on replace because the newly created shape has another id temporarily on `preExecute` context. Maybe this would be an possibility also for detaching / attaching elements?

Reply:
@nikku @pinussilvestrus Or, we can use Map here instead of Object for handledNodes variable. The problem with Objects that, their keys have to be String but Maps can have objects as key, so we can use directly the nodes as keys.

Check out this example:

```
var map = new Map();
var obj1 = {property1: true};
var obj2 = {property1: true};
map.set(obj1, "SOME VALUE");
map.has(obj2) // returns false
```
in this example obj1 and obj2 has exactly the same properties, values, but Map.has method returns false

Reply:
I think that's a good idea to ensure uniqueness. 👍 Nevertheless, it's interesting why we have the same id when detaching/attaching one element, and don't have the same behavior as in the replacement. We should maybe have a look at this in the future to prevent possible resulting bugs in other places, but for now, to fix this problem, your solution is good enough to fix the general uniqueness problem in the `UpdateFlowNodeRefsHandler` when we can't guarantee uniqueness with the ids.

Reply:
I ended up finding a better solution and get rid of handledNodes control variable, as this variable is only used to check if we have processed an element or not, so actually allFlowNodeShapes.indexOf(node) gives us the same result anyway and it also provides uniqueness without going into id/type.

Reply:
Did you think about the performance implications of not using a map? How many arrays are scanned how often for huge diagrams, i.e. [this one](https://demo.bpmn.io/s/application-processing).

Reply:
No I did not consider this, however we cannot use Map as they are not compatible with the JS version used in our test framework (PhantomJS).

As a performance remark:

```
var ary = [];
for (var i = 0; i < 1000000; i++){
	ary.push({testProp: true});
}
var obj = {testProp: true};
ary.push(obj); // put array to last as the worst case scenario

var startTime = performance.now();
ary.indexOf(obj);
console.log(performance.now() - startTime);  // 0.8 ms
```

So it takes 0.8 ms for indexOf to execute for 1 million + 1 nodes. I believe that amount of nodes are rarely the case and we have another problems caused by usage of document.elementFromPoint in Hover related classes if we happen to have a diagram with 1M+ nodes.

The comparison of objects are performed with (obj1 === obj), which is basically a reference comparison, so it's rather cheap.

Reply:
Thanks for the clarification. I did another test and verified we don't run in a `O^n` setting here.